File: java/src/test/java/br/com/vps/consulting/b2b/management/partner/infrastructure/persistence/jpa/DefaultPartnerRepositoryIT.java
package br.com.vps.consulting.b2b.management.partner.infrastructure.persistence.jpa;

import br.com.vps.consulting.b2b.management.TestcontainersConfiguration;
import br.com.vps.consulting.b2b.management.partner.domain.PartnerId;
import br.com.vps.consulting.b2b.management.partner.infrastructure.persistence.DefaultPartnerCreditRepository;
import br.com.vps.consulting.b2b.management.partner.infrastructure.persistence.PartnerCreditEntity;
import br.com.vps.consulting.b2b.management.partner.infrastructure.persistence.PartnerEntity;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.autoconfigure.ImportAutoConfiguration;
import org.springframework.boot.data.jpa.test.autoconfigure.DataJpaTest;
import org.springframework.boot.flyway.autoconfigure.FlywayAutoConfiguration;
import org.springframework.boot.jdbc.test.autoconfigure.AutoConfigureTestDatabase;
import org.springframework.context.annotation.Import;
import org.springframework.dao.DataIntegrityViolationException;

import java.math.BigDecimal;
import java.time.Instant;
import java.util.UUID;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

@DataJpaTest
@AutoConfigureTestDatabase(replace = AutoConfigureTestDatabase.Replace.NONE)
@Import({TestcontainersConfiguration.class, DefaultPartnerCreditRepository.class})
@ImportAutoConfiguration(FlywayAutoConfiguration.class)
class DefaultPartnerRepositoryIT {

    @Autowired private DefaultPartnerCreditRepository adapter;
    @Autowired private PartnerJpaRepository partnerJpaRepository;
    @Autowired private PartnerCreditJpaRepository partnerCreditJpaRepository;

    @Test
    @DisplayName("Given a sufficient free balance, when reserveCredit is called, should reserve the credit")
    void shouldReserveCredit_whenFreeBalanceSufficient() {
        final var partnerId = setup("1000.00", "0.00");

        final var result = adapter.reserveCredit(PartnerId.from(partnerId), new BigDecimal("300.00"));

        assertThat(result).isTrue();
        final var credit = findCredit(partnerId);
        assertThat(credit.getReservedBalance()).isEqualByComparingTo("300.00");
        assertThat(credit.getAvailableBalance()).isEqualByComparingTo("1000.00");
    }

    @Test
    @DisplayName("Given an amount that exactly equals the free balance, when reserveCredit is called, should reserve the credit")
    void shouldReserveCredit_whenAmountEqualsExactFreeBalance() {
        final var partnerId = setup("1000.00", "700.00");

        final var result = adapter.reserveCredit(PartnerId.from(partnerId), new BigDecimal("300.00"));

        assertThat(result).isTrue();
        final var credit = findCredit(partnerId);
        assertThat(credit.getReservedBalance()).isEqualByComparingTo("1000.00");
    }

    @Test
    @DisplayName("Given an insufficient free balance, when reserveCredit is called, should not reserve the credit")
    void shouldNotReserveCredit_whenFreeBalanceInsufficient() {
        final var partnerId = setup("1000.00", "800.00");

        final var result = adapter.reserveCredit(PartnerId.from(partnerId), new BigDecimal("300.00"));

        assertThat(result).isFalse();
        final var credit = findCredit(partnerId);
        assertThat(credit.getReservedBalance()).isEqualByComparingTo("800.00");
    }

    @Test
    @DisplayName("Given all balance already reserved, when reserveCredit is called, should not reserve any more credit")
    void shouldNotReserveCredit_whenNoFreeBalance() {
        final var partnerId = setup("500.00", "500.00");

        final var result = adapter.reserveCredit(PartnerId.from(partnerId), new BigDecimal("1.00"));

        assertThat(result).isFalse();
        assertThat(findCredit(partnerId).getReservedBalance()).isEqualByComparingTo("500.00");
    }

    @Test
    @DisplayName("Given a reserved balance, when debitReservation is called, should convert it into a real debit")
    void shouldDebitReservation_whenCalled() {
        final var partnerId = setup("1000.00", "300.00");

        adapter.debitReservation(PartnerId.from(partnerId), new BigDecimal("300.00"));

        final var credit = findCredit(partnerId);
        assertThat(credit.getAvailableBalance()).isEqualByComparingTo("700.00");
        assertThat(credit.getReservedBalance()).isEqualByComparingTo("0.00");
    }

    @Test
    @DisplayName("Given an amount above the available balance, when debitReservation is called, should throw DataIntegrityViolationException")
    void shouldThrow_whenDebitReservationViolatesConstraint() {
        final var partnerId = setup("100.00", "100.00");

        assertThatThrownBy(() -> adapter.debitReservation(PartnerId.from(partnerId), new BigDecimal("200.00")))
                .isInstanceOf(DataIntegrityViolationException.class);
    }

    // ── releaseReservation ───────────────────────────────────────────────────

    @Test
    @DisplayName("Given a reserved balance, when releaseReservation is called, should reduce it without touching the available balance")
    void shouldReleaseReservation_whenCalled() {
        final var partnerId = setup("1000.00", "300.00");

        adapter.releaseReservation(PartnerId.from(partnerId), new BigDecimal("300.00"));

        final var credit = findCredit(partnerId);
        assertThat(credit.getReservedBalance()).isEqualByComparingTo("0.00");
        assertThat(credit.getAvailableBalance()).isEqualByComparingTo("1000.00");
    }

    @Test
    @DisplayName("Given an amount above the reserved balance, when releaseReservation is called, should throw DataIntegrityViolationException")
    void shouldThrow_whenReleaseReservationViolatesConstraint() {
        final var partnerId = setup("1000.00", "100.00");

        assertThatThrownBy(() -> adapter.releaseReservation(PartnerId.from(partnerId), new BigDecimal("200.00")))
                .isInstanceOf(DataIntegrityViolationException.class);
    }

    // ── refundCredit ─────────────────────────────────────────────────────────

    @Test
    @DisplayName("Given a debited amount, when refundCredit is called, should restore the available balance")
    void shouldRefundCredit_whenCalled() {
        final var partnerId = setup("700.00", "0.00");

        adapter.refundCredit(PartnerId.from(partnerId), new BigDecimal("300.00"));

        final var credit = findCredit(partnerId);
        assertThat(credit.getAvailableBalance()).isEqualByComparingTo("1000.00");
        assertThat(credit.getReservedBalance()).isEqualByComparingTo("0.00");
    }

    // ── Ciclos completos ──────────────────────────────────────────────────────

    @Test
    @DisplayName("Given a reserve followed by a debit, when the create-to-approve cycle completes, should leave the net credit consumed")
    void shouldCompleteCycle_createThenApprove() {
        final var partnerId = setup("1000.00", "0.00");
        final var pid = PartnerId.from(partnerId);

        adapter.reserveCredit(pid, new BigDecimal("300.00"));
        adapter.debitReservation(pid, new BigDecimal("300.00"));

        final var credit = findCredit(partnerId);
        assertThat(credit.getAvailableBalance()).isEqualByComparingTo("700.00");
        assertThat(credit.getReservedBalance()).isEqualByComparingTo("0.00");
    }

    @Test
    @DisplayName("Given a reserve followed by a release, when the create-to-cancel-pending cycle completes, should leave the balance unchanged")
    void shouldCompleteCycle_createThenCancelPending() {
        final var partnerId = setup("1000.00", "0.00");
        final var pid = PartnerId.from(partnerId);

        adapter.reserveCredit(pid, new BigDecimal("300.00"));
        adapter.releaseReservation(pid, new BigDecimal("300.00"));

        final var credit = findCredit(partnerId);
        assertThat(credit.getAvailableBalance()).isEqualByComparingTo("1000.00");
        assertThat(credit.getReservedBalance()).isEqualByComparingTo("0.00");
    }

    @Test
    @DisplayName("Given a reserve, debit and refund, when the create-to-approve-to-cancel cycle completes, should leave the balance unchanged")
    void shouldCompleteCycle_createApproveCancel() {
        final var partnerId = setup("1000.00", "0.00");
        final var pid = PartnerId.from(partnerId);

        adapter.reserveCredit(pid, new BigDecimal("300.00"));
        adapter.debitReservation(pid, new BigDecimal("300.00"));
        adapter.refundCredit(pid, new BigDecimal("300.00"));

        final var credit = findCredit(partnerId);
        assertThat(credit.getAvailableBalance()).isEqualByComparingTo("1000.00");
        assertThat(credit.getReservedBalance()).isEqualByComparingTo("0.00");
    }

    private UUID setup(final String availableBalance, final String reservedBalance) {
        final var partnerId = UUID.randomUUID();
        partnerJpaRepository.save(PartnerEntity.builder()
                .id(partnerId)
                .name("Test Partner")
                .document(UUID.randomUUID().toString().replace("-", "").substring(0, 14))
                .createdAt(Instant.now())
                .build());
        partnerCreditJpaRepository.save(PartnerCreditEntity.builder()
                .id(partnerId)
                .creditLimit(new BigDecimal("1000.00"))
                .availableBalance(new BigDecimal(availableBalance))
                .reservedBalance(new BigDecimal(reservedBalance))
                .updatedAt(Instant.now())
                .build());
        return partnerId;
    }

    private PartnerCreditEntity findCredit(final UUID partnerId) {
        return partnerCreditJpaRepository.findById(partnerId).orElseThrow();
    }

}

File: kotlin/src/test/kotlin/br/com/vps/consulting/b2b/management/partner/infrastructure/persistence/jpa/DefaultPartnerRepositoryIT.kt
package br.com.vps.consulting.b2b.management.partner.infrastructure.persistence.jpa

import br.com.vps.consulting.b2b.management.TestcontainersConfiguration
import br.com.vps.consulting.b2b.management.partner.domain.Partner
import br.com.vps.consulting.b2b.management.partner.domain.PartnerId
import br.com.vps.consulting.b2b.management.partner.infrastructure.persistence.DefaultPartnerRepository
import br.com.vps.consulting.b2b.management.shared.core.vo.Money
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.autoconfigure.ImportAutoConfiguration
import org.springframework.boot.data.jpa.test.autoconfigure.DataJpaTest
import org.springframework.boot.flyway.autoconfigure.FlywayAutoConfiguration
import org.springframework.boot.jdbc.test.autoconfigure.AutoConfigureTestDatabase
import org.springframework.context.annotation.Import
import java.math.BigDecimal
import java.util.UUID

@DataJpaTest
@AutoConfigureTestDatabase(replace = AutoConfigureTestDatabase.Replace.NONE)
@Import(TestcontainersConfiguration::class, DefaultPartnerRepository::class)
@ImportAutoConfiguration(FlywayAutoConfiguration::class)
class DefaultPartnerRepositoryIT {

    @Autowired
    lateinit var adapter: DefaultPartnerRepository

    @Autowired
    lateinit var partnerJpaRepository: PartnerJpaRepository

    @Autowired
    lateinit var partnerBalanceJpaRepository: PartnerBalanceJpaRepository

    private fun newPartner() = Partner.create(name = "Acme Corp", document = UUID.randomUUID().toString().replace("-", "").take(14))

    @Test
    fun `should save partner and create a zero-balance row`() {
        val partner = newPartner()

        adapter.save(partner)

        assertThat(partnerJpaRepository.findById(partner.id.value)).isPresent
        val balance = partnerBalanceJpaRepository.findById(partner.id.value).orElseThrow()
        assertThat(balance.totalBalance).isEqualByComparingTo(BigDecimal.ZERO)
        assertThat(balance.availableBalance).isEqualByComparingTo(BigDecimal.ZERO)
    }

    @Test
    fun `should find partner by id`() {
        val partner = newPartner()
        adapter.save(partner)

        val found = adapter.findById(partner.id)

        assertThat(found).isNotNull
        assertThat(found!!.document).isEqualTo(partner.document)
    }

    @Test
    fun `should return null when partner not found by id`() {
        assertThat(adapter.findById(PartnerId.generate())).isNull()
    }

    @Test
    fun `should find balance by id returning PartnerBalance with matching values`() {
        val partner = newPartner()
        adapter.save(partner)
        adapter.creditBalance(partner.id, Money.of(BigDecimal("100.00")))

        val balance = adapter.findBalanceById(partner.id)

        assertThat(balance).isNotNull
        assertThat(balance!!.totalBalance.value).isEqualByComparingTo("100.00")
        assertThat(balance.availableBalance.value).isEqualByComparingTo("100.00")
    }

    @Test
    fun `should return null when balance not found by id`() {
        assertThat(adapter.findBalanceById(PartnerId.generate())).isNull()
    }

    @Test
    fun `should findAll with no document filter returning all partners ordered by createdAt desc`() {
        val first = newPartner()
        adapter.save(first)
        val second = newPartner()
        adapter.save(second)

        val page = adapter.findAll(document = null, pageSize = 20, pageNumber = 0)

        assertThat(page.items.map { it.id.value }).contains(first.id.value, second.id.value)
    }

    @Test
    fun `should findAll filtered by document returning only matching partner`() {
        val partner = newPartner()
        adapter.save(partner)
        adapter.save(newPartner())

        val page = adapter.findAll(document = partner.document, pageSize = 20, pageNumber = 0)

        assertThat(page.items).hasSize(1)
        assertThat(page.items[0].document).isEqualTo(partner.document)
    }

    @Test
    fun `should findAll respecting pageSize and pageNumber`() {
        repeat(3) { adapter.save(newPartner()) }

        val page = adapter.findAll(document = null, pageSize = 2, pageNumber = 0)

        assertThat(page.items).hasSize(2)
        assertThat(page.totalElements).isGreaterThanOrEqualTo(3L)
    }

    @Test
    fun `should creditBalance increase both totalBalance and availableBalance and return true`() {
        val partner = newPartner()
        adapter.save(partner)

        val result = adapter.creditBalance(partner.id, Money.of(BigDecimal("50.00")))

        assertThat(result).isTrue()
        val balance = adapter.findBalanceById(partner.id)!!
        assertThat(balance.totalBalance.value).isEqualByComparingTo("50.00")
        assertThat(balance.availableBalance.value).isEqualByComparingTo("50.00")
    }

    @Test
    fun `should debitBalance decrease availableBalance only and return true when sufficient balance`() {
        val partner = newPartner()
        adapter.save(partner)
        adapter.creditBalance(partner.id, Money.of(BigDecimal("100.00")))

        val result = adapter.debitBalance(partner.id, Money.of(BigDecimal("30.00")))

        assertThat(result).isTrue()
        val balance = adapter.findBalanceById(partner.id)!!
        assertThat(balance.availableBalance.value).isEqualByComparingTo("70.00")
        assertThat(balance.totalBalance.value).isEqualByComparingTo("100.00")
    }

    @Test
    fun `should debitBalance return false and leave balance unchanged when insufficient balance`() {
        val partner = newPartner()
        adapter.save(partner)
        adapter.creditBalance(partner.id, Money.of(BigDecimal("10.00")))

        val result = adapter.debitBalance(partner.id, Money.of(BigDecimal("50.00")))

        assertThat(result).isFalse()
        val balance = adapter.findBalanceById(partner.id)!!
        assertThat(balance.availableBalance.value).isEqualByComparingTo("10.00")
    }

    @Test
    fun `should creditBalance return false when partner balance row does not exist`() {
        val result = adapter.creditBalance(PartnerId.generate(), Money.of(BigDecimal("10.00")))

        assertThat(result).isFalse()
    }

}
